heron/proto/ckptmgr.proto

```diff
@@ -0,0 +1,182 @@
+  optional string oldest_checkpoint_preserved = 1;
+  optional bool clean_all_checkpoints = 2 [default = false];
+}
+
+// Message sent by ckptmgr to tmaster about the cleanup request
+message CleanStatefulCheckpointResponse {
+  required heron.proto.system.Status status = 1;
+}
+
+// stmgr -> ckptmgr messages
+
+// This is the message that a stmgr sends
+// when it wants to establish the connection
+// with checkpoint manager
+message RegisterStMgrRequest{
+  required string topology_name = 1;
+  required string topology_id = 2;
+  required string stmgr = 3;
+}
+
+// This is the message that checkpoint manager
+// sends when it receives the register request
+// from stmgr
+message RegisterStMgrResponse {
+  required heron.proto.system.Status status = 1;
+}
+
+// This is the request that StMgr sends to Checkpoint Mgr
+// to store instance checkpoint data
+message SaveInstanceStateRequest {
+  // Information about the instance whose state this is
+  required heron.proto.system.Instance instance = 1;
+  required InstanceStateCheckpoint checkpoint = 2;
+}
+
+// After successfully saving state, this is what
+// Checkpoint Mgr sends back as response
+message SaveInstanceStateResponse {
+  required heron.proto.system.Status status = 1;
+  required heron.proto.system.Instance instance = 2;
+  required string checkpoint_id = 3;
+}
+
+// This is the request that StMgr sends to Checkpoint Mgr
+// to fetch the instance state data
+message GetInstanceStateRequest {
+  required heron.proto.system.Instance instance = 1;
+  required string checkpoint_id = 2;
+}
+
+// This is the response that Checkpoint Mgr sends to StMgr
+// for its get instance state request
+message GetInstanceStateResponse {
+  required heron.proto.system.Status status = 1;
+  required heron.proto.system.Instance instance = 2;
+  required string checkpoint_id = 3;
+  optional InstanceStateCheckpoint checkpoint = 4;
+}
+
+// stmgr -> Instance messages
+
+// This is the message that the stmgr sends to its
+// local tasks to begin initiating checkpoint
+message InitiateStatefulCheckpoint {
+  required string checkpoint_id = 1;
+}
+
+// This is the message that the instance sends
+// when it wants stmgr to store its state
+message InstanceStateCheckpoint {
+  required string checkpoint_id = 1;
+  required bytes state = 2;
+}
+
+// This is the message that stmgr sends to its instace
+// asking them to restore their state
+message RestoreInstanceStateRequest {
+  required InstanceStateCheckpoint state = 1;
+}
+
+// This is the message that instance sends to its stmgr
+// when done with restoring the state
+message RestoreInstanceStateResponse {
+  required string checkpoint_id = 1;
+}
+
+// Message sent by stmgr to Instance to start processing
+message StartInstanceStatefulProcessing {
+  required string checkpoint_id = 1;
+}
+
+// stmgr -> stmgr messages
+
+// This is the message that stmgr sends to other stmgr
+// to send it to downstream task
+message DownstreamStatefulCheckpoint {
```

Comment:
is this message to handle the case where a downstream instance has multiple updtream instances?